In [1]:
#  INSTALL IPYTHON-SQL:
# !pip install ipython-sql --quiet


# IMPORT NECESSARY LIBRARIES:
import pandas as pd
import sqlite3


# DEFINE PATHS:
DF_PATH = './data/BankChurners.csv'
DB_PATH = './data/bank_churners.db'

Reading the data into a pandas a dataframe: 

In [2]:
df = pd.read_csv(DF_PATH)
df.head()

CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0  768805383  Existing Customer            45      M                3   
1  818770008  Existing Customer            49      F                5   
2  713982108  Existing Customer            51      M                3   
3  769911858  Existing Customer            40      F                4   
4  709106358  Existing Customer            40      M                3   

  Education_Level Marital_Status Income_Category Card_Category  \
0     High School        Married     $60K - $80K          Blue   
1        Graduate         Single  Less than $40K          Blue   
2        Graduate        Married    $80K - $120K          Blue   
3     High School        Unknown  Less than $40K          Blue   
4      Uneducated        Married     $60K - $80K          Blue   

   Months_on_book  ...  Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  \
0              39  ...       12691.0                  777          11914.0   
1              44  ...        8256.0                  864           7392.0   
2              36  ...        3418.0                    0           3418.0   
3              34  ...        3313.0                 2517            796.0   
4              21  ...        4716.0                    0           4716.0   

   Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
0                 1.335             1144              42                1.625   
1                 1.541             1291              33                3.714   
2                 2.594             1887              20                2.333   
3                 1.405             1171              20                2.333   
4                 2.175              816              28                2.500   

   Avg_Utilization_Ratio  \
0                  0.061   
1                  0.105   
2                  0.000   
3                  0.760   
4                  0.000   

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_1  \
0                                           0.000093       
1                                           0.000057       
2                                           0.000021       
3                                           0.000134       
4                                           0.000022       

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_2  
0                                            0.99991      
1                                            0.99994      
2                                            0.99998      
3                                            0.99987      
4                                            0.99998      

[5 rows x 23 columns]

In [3]:
cnn = sqlite3.connect(DB_PATH)
df.to_sql('customers', cnn)

10127

In [4]:
%load_ext sql
%sql sqlite:///data/bank_churners.db 

In [6]:
%%sql

SELECT * 
from customers
where ROW_NUM < 5

 * sqlite:///data/bank_churners.db
(sqlite3.OperationalError) no such column: ROW_NUM
[SQL: SELECT * 
from customers
where ROW_NUM < 5]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


In [ ]:
result = _ 
result

DESCRIBE THE TABLE:

In [9]:
%%sql 

SELECT *
FROM customers;
PRAGMA table_info(customers);

 * sqlite:///data/bank_churners.db
Done.
Done.


[(0, 'index', 'INTEGER', 0, None, 0),
 (1, 'CLIENTNUM', 'INTEGER', 0, None, 0),
 (2, 'Attrition_Flag', 'TEXT', 0, None, 0),
 (3, 'Customer_Age', 'INTEGER', 0, None, 0),
 (4, 'Gender', 'TEXT', 0, None, 0),
 (5, 'Dependent_count', 'INTEGER', 0, None, 0),
 (6, 'Education_Level', 'TEXT', 0, None, 0),
 (7, 'Marital_Status', 'TEXT', 0, None, 0),
 (8, 'Income_Category', 'TEXT', 0, None, 0),
 (9, 'Card_Category', 'TEXT', 0, None, 0),
 (10, 'Months_on_book', 'INTEGER', 0, None, 0),
 (11, 'Total_Relationship_Count', 'INTEGER', 0, None, 0),
 (12, 'Months_Inactive_12_mon', 'INTEGER', 0, None, 0),
 (13, 'Contacts_Count_12_mon', 'INTEGER', 0, None, 0),
 (14, 'Credit_Limit', 'REAL', 0, None, 0),
 (15, 'Total_Revolving_Bal', 'INTEGER', 0, None, 0),
 (16, 'Avg_Open_To_Buy', 'REAL', 0, None, 0),
 (17, 'Total_Amt_Chng_Q4_Q1', 'REAL', 0, None, 0),
 (18, 'Total_Trans_Amt', 'INTEGER', 0, None, 0),
 (19, 'Total_Trans_Ct', 'INTEGER', 0, None, 0),
 (20, 'Total_Ct_Chng_Q4_Q1', 'REAL', 0, None, 0),
 (21, 'Avg_Utilization_Ratio', 'REAL', 0, None, 0),
 (22, 'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_1', 'REAL', 0, None, 0),
 (23, 'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_2', 'REAL', 0, None, 0)]

/*data cleaning*/
-- checking for misssing values

In [11]:
%%sql

SELECT *
FROM customers
WHERE CLIENTNUM is NULL OR Attrition_Flag IS NULL OR Customer_Age IS NULL OR Gender IS NULL OR Dependent_count is NULL OR Education_Level IS NULL OR Marital_Status IS Null;

 * sqlite:///data/bank_churners.db
Done.


[]

-- there are no missing values

/*unique Customers
There's a total of 10127 customers*/

In [13]:
%%sql 

SELECT 
DISTINCT CLIENTNUM
FROM customers
LIMIT 5;

 * sqlite:///data/bank_churners.db
Done.


[(768805383,), (818770008,), (713982108,), (769911858,), (709106358,)]

In [14]:
%%sql 

SELECT 
concat(round(COUNT(CASE WHEN Attrition_Flag = '1' THEN Clientnum END) / COUNT(Clientnum),2) * 100,'%') AS churn_rate,
concat(round(COUNT(CASE WHEN Attrition_Flag = '0' THEN Clientnum END) / COUNT(Clientnum),2) * 100,'%') AS retention_rate
FROM customers;

 * sqlite:///data/bank_churners.db
(sqlite3.OperationalError) no such function: concat
[SQL: SELECT 
concat(round(COUNT(CASE WHEN Attrition_Flag = '1' THEN Clientnum END) / COUNT(Clientnum),2) * 100,'%') AS churn_rate,
concat(round(COUNT(CASE WHEN Attrition_Flag = '0' THEN Clientnum END) / COUNT(Clientnum),2) * 100,'%') AS retention_rate
FROM customers;]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


In [21]:
%%sql 

SELECT DISTINCT Attrition_Flag
FROM customers;

 * sqlite:///data/bank_churners.db
Done.


[('Existing Customer',), ('Attrited Customer',)]

/*Column Manipulation
transforming the customer_age column into bins using the case statement*/

In [24]:
%%sql

SELECT 
COUNT(Attrition_FLag) AS existing_customer_count
FROM customers
WHERE Attrition_Flag ='Existing Customer';

 * sqlite:///data/bank_churners.db
Done.


[(8500,)]

-- there are 8500 existing customers

In [25]:
%%sql

SELECT 
COUNT(Attrition_Flag) AS churned_customer_count
FROM customers
WHERE Attrition_Flag ='Attrited Customer';

 * sqlite:///data/bank_churners.db
Done.


[(1627,)]

-- there are 1627 churned customers indicating high imbalance

In [27]:
%%sql

SELECT
Attrition_FLag,
CASE 
WHEN Attrition_FLag ='Attrited Customer' THEN 1 
ELSE 0 END AS attrited_customer
FROM customers
GROUP BY Attrition_Flag;

 * sqlite:///data/bank_churners.db
Done.


[('Attrited Customer', 1), ('Existing Customer', 0)]

In [28]:
%%sql

UPDATE customers
SET Attrition_Flag=CASE 
WHEN Attrition_FLag ='Attrited Customer' THEN 1 
ELSE 0 END;
SELECT 
Attrition_Flag
FROM customers
GROUP BY Attrition_Flag;

 * sqlite:///data/bank_churners.db
10127 rows affected.
Done.


[('0',), ('1',)]

In [30]:
%%sql

SELECT 
round(COUNT(CASE WHEN Attrition_Flag = 1 THEN CLIENTNUM END) / COUNT(CLIENTNUM), 2) * 100  AS churn_rate,
round(COUNT(CASE WHEN Attrition_Flag = 0 THEN CLIENTNUM END) / COUNT(CLIENTNUM), 2) * 100  AS retention_rate
FROM customers;

 * sqlite:///data/bank_churners.db
Done.


[(0.0, 0.0)]

-- customer Age

In [31]:
%%sql 

SELECT 
MAX(Customer_Age) AS max_age,
MIN(Customer_Age) AS min_age
FROM customers;

 * sqlite:///data/bank_churners.db
Done.


[(73, 26)]

-- the eldest customer is 73 while the youngest is 26

In [33]:
%%sql 

SELECT 
    Customer_Age,
    SUM(CASE WHEN Attrition_Flag = '1' THEN 1 ELSE 0 END) AS churned_customer_count,
    SUM(CASE WHEN Attrition_Flag = '0' THEN 1 ELSE 0 END) AS existing_customer_count,
    ROUND((SUM(CASE WHEN Attrition_Flag = '1' THEN 1 ELSE 0 END) * 1.0 / (SUM(CASE WHEN Attrition_Flag = '0' THEN 1 ELSE 0 END) + SUM(CASE WHEN Attrition_Flag = '1' THEN 1 ELSE 0 END))) * 100, 2) || '%' AS churn_rate,
    ROUND((SUM(CASE WHEN Attrition_Flag = '0' THEN 1 ELSE 0 END) * 1.0 / (SUM(CASE WHEN Attrition_Flag = '1' THEN 1 ELSE 0 END) + SUM(CASE WHEN Attrition_Flag = '0' THEN 1 ELSE 0 END))) * 100, 2) || '%' AS retention_rate
FROM customers
GROUP BY Customer_Age
ORDER BY Customer_Age ASC;

 * sqlite:///data/bank_churners.db
Done.


[(26, 6, 72, '7.69%', '92.31%'),
 (27, 3, 29, '9.38%', '90.63%'),
 (28, 1, 28, '3.45%', '96.55%'),
 (29, 7, 49, '12.5%', '87.5%'),
 (30, 15, 55, '21.43%', '78.57%'),
 (31, 13, 78, '14.29%', '85.71%'),
 (32, 17, 89, '16.04%', '83.96%'),
 (33, 20, 107, '15.75%', '84.25%'),
 (34, 19, 127, '13.01%', '86.99%'),
 (35, 21, 163, '11.41%', '88.59%'),
 (36, 24, 197, '10.86%', '89.14%'),
 (37, 37, 223, '14.23%', '85.77%'),
 (38, 47, 256, '15.51%', '84.49%'),
 (39, 48, 285, '14.41%', '85.59%'),
 (40, 64, 297, '17.73%', '82.27%'),
 (41, 76, 303, '20.05%', '79.95%'),
 (42, 62, 364, '14.55%', '85.45%'),
 (43, 85, 388, '17.97%', '82.03%'),
 (44, 84, 416, '16.8%', '83.2%'),
 (45, 79, 407, '16.26%', '83.74%'),
 (46, 82, 408, '16.73%', '83.27%'),
 (47, 76, 403, '15.87%', '84.13%'),
 (48, 85, 387, '18.01%', '81.99%'),
 (49, 79, 416, '15.96%', '84.04%'),
 (50, 71, 381, '15.71%', '84.29%'),
 (51, 58, 340, '14.57%', '85.43%'),
 (52, 58, 318, '15.43%', '84.57%'),
 (53, 59, 328, '15.25%', '84.75%'),
 (54, 69, 238, '22.48%', '77.52%'),
 (55, 51, 228, '18.28%', '81.72%'),
 (56, 43, 219, '16.41%', '83.59%'),
 (57, 33, 190, '14.8%', '85.2%'),
 (58, 24, 133, '15.29%', '84.71%'),
 (59, 40, 117, '25.48%', '74.52%'),
 (60, 13, 114, '10.24%', '89.76%'),
 (61, 17, 76, '18.28%', '81.72%'),
 (62, 17, 76, '18.28%', '81.72%'),
 (63, 8, 57, '12.31%', '87.69%'),
 (64, 5, 38, '11.63%', '88.37%'),
 (65, 9, 92, '8.91%', '91.09%'),
 (66, 1, 1, '50.0%', '50.0%'),
 (67, 0, 4, '0.0%', '100.0%'),
 (68, 1, 1, '50.0%', '50.0%'),
 (70, 0, 1, '0.0%', '100.0%'),
 (73, 0, 1, '0.0%', '100.0%')]

In [35]:
%%sql

SELECT 
    Gender,
    SUM(CASE WHEN Attrition_Flag = '1' THEN 1 ELSE 0 END) AS churned_customer_count,
    SUM(CASE WHEN Attrition_Flag = '0' THEN 1 ELSE 0 END) AS existing_customer_count,
    CAST(ROUND((SUM(CASE WHEN Attrition_Flag = '1' THEN 1 ELSE 0 END) * 1.0 / (SUM(CASE WHEN Attrition_Flag = '0' THEN 1 ELSE 0 END) + SUM(CASE WHEN Attrition_Flag = '1' THEN 1 ELSE 0 END))) * 100, 2) AS TEXT) || '%' AS churn_rate
FROM customers
GROUP BY Gender
ORDER BY Gender ASC;


 * sqlite:///data/bank_churners.db
Done.


[('F', 930, 4428, '17.36%'), ('M', 697, 4072, '14.62%')]

In [36]:
%%sql 

SELECT 
Dependent_count,
SUM(CASE WHEN Attrition_Flag = '1' THEN 1 ELSE 0 END) AS churned_customer_count,
SUM(CASE WHEN Attrition_Flag = '0' THEN 1 ELSE 0 END) AS existing_customer_count,
ROUND((SUM(CASE WHEN Attrition_Flag = '1' THEN 1 ELSE 0 END) / (SUM(CASE WHEN Attrition_Flag = '0' THEN 1 ELSE 0 END) + SUM(CASE WHEN Attrition_Flag = '1' THEN 1 ELSE 0 END))) * 100, 2) || '%' AS churn_rate
FROM customers
GROUP BY Dependent_count
ORDER BY Dependent_count ASC; 

 * sqlite:///data/bank_churners.db
Done.


[(0, 135, 769, '0.0%'),
 (1, 269, 1569, '0.0%'),
 (2, 417, 2238, '0.0%'),
 (3, 482, 2250, '0.0%'),
 (4, 260, 1314, '0.0%'),
 (5, 64, 360, '0.0%')]